plume-models/src/timeline/mod.rs

```diff
@@ -0,0 +1,845 @@
+                        .or(timeline_definition::user_id.is_null()),
+                )
+                .load::<Self>(conn)
+                .map_err(Error::from)
+        } else {
+            timeline_definition::table
+                .filter(timeline_definition::user_id.is_null())
+                .load::<Self>(conn)
+                .map_err(Error::from)
+        }
+    }
+
+    pub fn new_for_user(
+        conn: &Connection,
+        user_id: i32,
+        name: String,
+        query_string: String,
+    ) -> Result<Timeline> {
+        {
+            let query = TimelineQuery::parse(&query_string)?; // verify the query is valid
+            if let Some(err) =
+                query
+                    .list_used_lists()
+                    .into_iter()
+                    .find_map(|(name, kind)| {
+                        let list = List::find_for_user_by_name(conn, Some(user_id), &name)
+                            .map(|l| l.kind() == kind);
+                        match list {
+                            Ok(true) => None,
+                            Ok(false) => Some(Error::TimelineQuery(QueryError::RuntimeError(
+                                format!("list '{}' has the wrong type for this usage", name),
+                            ))),
+                            Err(_) => Some(Error::TimelineQuery(QueryError::RuntimeError(
+                                format!("list '{}' was not found", name),
+                            ))),
+                        }
+                    })
+            {
+                Err(err)?;
+            }
+        }
+        Self::insert(
+            conn,
+            NewTimeline {
+                user_id: Some(user_id),
+                name,
+                query: query_string,
+            },
+        )
+    }
+
+    pub fn new_for_instance(
+        conn: &Connection,
+        name: String,
+        query_string: String,
+    ) -> Result<Timeline> {
+        {
+            let query = TimelineQuery::parse(&query_string)?; // verify the query is valid
+            if let Some(err) =
+                query
+                    .list_used_lists()
+                    .into_iter()
+                    .find_map(|(name, kind)| {
+                        let list = List::find_for_user_by_name(conn, None, &name)
+                            .map(|l| l.kind() == kind);
+                        match list {
+                            Ok(true) => None,
+                            Ok(false) => Some(Error::TimelineQuery(QueryError::RuntimeError(
+                                format!("list '{}' has the wrong type for this usage", name),
+                            ))),
+                            Err(_) => Some(Error::TimelineQuery(QueryError::RuntimeError(
+                                format!("list '{}' was not found", name),
+                            ))),
+                        }
+                    })
+            {
+                Err(err)?;
```

Comment:
that's another 21 lines of error handling

Reply:
there should be a better way to do that …
(I'm not saying you have to find that *now*)